### PR TITLE
Implement RegisterMapInterfaceINTEL execution mode

### PR DESCRIPTION
**Pull request by tiwaria1**
This PR adds the new execution mode `RegisterMapInterfaceINTEL`, see the Khronos
SPIRV spec here: https://github.com/KhronosGroup/SPIRV-Registry/pull/176
This execution mode allows specifying a 'register' based interface for FPGA kernels.

The RegisterMapInterfaceINTEL execution mode is added with a 0/1 literal based on
the kernel metadata. When the metadata is:
```
!ip_interface !N
!N = !{!"csr"}
```
The translator emits `RegisterMapInterfaceINTEL 0`, and when the metadata is:
```
!ip_interface !N
!N = !{!"csr", !"accept_downstream_stall"}
```
The translator emits `RegisterMapInterfaceINTEL 1`

#### NOTE
- The new mode is under capability `FPGAKernelAttributesv2INTEL` which implicitly defines the capability `FPGAKernelAttributesv2INTEL`.
- This PR is very similar to https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1218

### CODE REVIEW

**Comment by tiwaria1** [comment]
@MrSidims Could you please help me review this as well? The tricky part here is that:
1. This exec mode is under capability FPGAKernelAttributesINTELv2 which implicitly defines the existing capability FPGAKernelAttributesINTEL. Not sure how exactly to reflect that in the code yet.
2. The spec states only one of StreamingInterfaceINTEL and RegisterMapInterfaceINTEL are valid at a time - should this be checked here in the translator or at some other level of the tool?

File: lib/SPIRV/SPIRVWriter.cpp
Comment (MrSidims) [inline lib/SPIRV/SPIRVWriter.cpp:4856]: AFAIK the translator isn't capable to track implicitly declared capabilities, it does look like a good enhancement for us. Like you did it here is LGTM

**Comment by MrSidims** [comment]
> The spec states only one of StreamingInterfaceINTEL and RegisterMapInterfaceINTEL are valid at a time - should this be checked here in the translator or at some other level of the tool?

We usually add checks on the translation level, see Unroll vs DontUnroll masks as an example. I don't remember, if it's a case for FPGA extensions, since we were relying on clang to check correctness of attributes back then (probably we still have some checks here as well), but it seems to be 100% idea now, since we no longer adding attributes to clang.

**Comment by MrSidims** [comment]
@tiwaria1 please don't forget to add a test :)

**Comment by tiwaria1** [comment]
> @tiwaria1 please don't forget to add a test :)
Added one. I did not run it locally though. Is there any automatic git CI testing that will run it for me on this PR?

**Comment by MrSidims** [comment]
> > @tiwaria1 please don't forget to add a test :)
> > Added one. I did not run it locally though. Is there any automatic git CI testing that will run it for me on this PR?

Yet here is, but I expect it to be broken right now due to LLVM upgrade to 17.0.0 . So I would suggest running tests locally (it won't faster the merge, since we would need to wait for the fixed CI, but still help with finding problems if any). To run test you should add `-DLLVM_SPIRV_INCLUDE_TESTS=ON` to cmake and run `make test` -j...

File: lib/SPIRV/libSPIRV/spirv_internal.hpp
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/spirv_internal.hpp:162]: Please restore an empty line

File: lib/SPIRV/SPIRVWriter.cpp
Comment (MrSidims) [inline lib/SPIRV/SPIRVWriter.cpp:4928]: Not quite correct, `CapabilityFPGAKernelAttributesv2INTEL` is needed only for `RegisterMapInterfaceINTEL` and not for other execution modes.

**Review by MrSidims — COMMENTED** [review]
To incorporate changes from SPIR-V headers please update spirv-headers-tag.conf

File: lib/SPIRV/SPIRVWriter.cpp
Comment (tiwaria1) [reply lib/SPIRV/SPIRVWriter.cpp:4928]: Are you suggesting I change the comment here or the code?

Comment (MrSidims) [inline lib/SPIRV/SPIRVWriter.cpp:4928]: The code like this or by splitting switch case:
```suggestion
        if (EMode == spv::ExecutionModeRegisterMapInterfaceINTEL)
          BM->addCapability(CapabilityFPGAKernelAttributesv2INTEL);
```
(note, the suggestion might be no clang-format friendly).

Comment (MrSidims) [reply lib/SPIRV/SPIRVWriter.cpp:4928]: The code, see below

Comment (tiwaria1) [reply lib/SPIRV/SPIRVWriter.cpp:4928]: Wait isn't this switch case already a switch on 'EMode'? So if control is here then EMode is definitely RegisterMapInterfaceINTEL so isn't this if condition trivially true always?

File: lib/SPIRV/libSPIRV/spirv_internal.hpp
Comment (tiwaria1) [reply lib/SPIRV/libSPIRV/spirv_internal.hpp]: Done

**Comment by tiwaria1** [comment]
> To incorporate changes from SPIR-V headers please update spirv-headers-tag.conf

I can only do that once the SPIRV-Headers PR is merged right?

File: lib/SPIRV/SPIRVWriter.cpp
Comment (MrSidims) [reply lib/SPIRV/SPIRVWriter.cpp:4928]: Like this `CapabilityFPGAKernelAttributesv2INTEL` will be emitted for for example `ExecutionModeMaxWorkDimINTEL` as well. If I read the spec correctly, it shouldn't be the case.

**Comment by MrSidims** [comment]
> > To incorporate changes from SPIR-V headers please update spirv-headers-tag.conf
> 
> I can only do that once the SPIRV-Headers PR is merged right?

Ah, sorry, I've seen notification, that the PR got closed and assumed, that is was merged.

File: lib/SPIRV/SPIRVWriter.cpp
Comment (tiwaria1) [reply lib/SPIRV/SPIRVWriter.cpp:4928]: You are right, I missed the fall-through in the cases above. Thanks for catching this!

**Review by MrSidims — COMMENTED** [review]
LGTM, just waiting for the headers be merged
Thanks!

**Comment by tiwaria1** [comment]
Hey @MrSidims the headers change has been merged. Has the CI been fixed yet?

**Review by MrSidims — APPROVED** [review]

**Comment by MrSidims** [comment]
@tiwaria1 please take a look at the test failure

**Comment by tiwaria1** [comment]
Having trouble building the translator locally, I did the following (llvm in-tree build steps):
```
git clone https://github.com/llvm/llvm-project.git
cd llvm-project/llvm/projects/
git clone https://github.com/tiwaria1/SPIRV-LLVM-Translator.git
# checkout my branch
mkdir build
cd build/
cmake ../llvm -DLLVM_ENABLE_PROJECTS="clang" -DCMAKE_BUILD_TYPE=Debug
make llvm-spirv -j`nproc`
```
and got the error:
```
/p/psg/swip/w/tiwaria1/ipauth/tiwaria1_specs/llvm-project/llvm/projects/SPIRV-LLVM-Translator/lib/SPIRV/SPIRVLowerConstExpr.cpp:48:10: fatal error: llvm/ADT/Triple.h: No such file or directory
 #include "llvm/ADT/Triple.h"
          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
projects/SPIRV-LLVM-Translator/lib/SPIRV/CMakeFiles/LLVMSPIRVLib.dir/build.make:257: recipe for target 'projects/SPIRV-LLVM-Translator/lib/SPIRV/CMakeFiles/LLVMSPIRVLib.dir/SPIRVLowerConstExpr.cpp.o' failed
```

**Comment by tiwaria1** [comment]
Not sure what's up with clang-format check, it says:
```
Run cat diff-to-inspect.txt | /usr/share/clang/clang-format-17/clang-format-diff.py -p1 > clang-format.patch
Traceback (most recent call last):
  File "/usr/share/clang/clang-format-17/clang-format-diff.py", line 177, in <module>
    main()
  File "/usr/share/clang/clang-format-17/clang-format-diff.py", line [13](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/actions/runs/4345812544/jobs/7591095087#step:6:14)5, in main
    process_subprocess_result(proc, args)
  File "/usr/share/clang/clang-format-[17](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/actions/runs/4345812544/jobs/7591095087#step:6:18)/clang-format-diff.py", line 42, in process_subprocess_result
    with open(filename) as f:
NameError: name 'filename' is not defined
Error: Process completed with exit code 1.
```

**Comment by MrSidims** [comment]
>> Not sure what's up with clang-format check, it says:

I'll take a look at clang-format check, you may ignore it
Please take a look at the failed test:
Failed Tests (1):
  LLVM_SPIRV :: extensions/INTEL/SPV_INTEL_kernel_attributes/register_map_interface_attribute.ll

```
/home/runner/work/SPIRV-LLVM-Translator/SPIRV-LLVM-Translator/SPIRV-LLVM-Translator/test/extensions/INTEL/SPV_INTEL_kernel_attributes/register_map_interface_attribute.ll:11:16: error: CHECK-SPIRV: expected string not found in input
; CHECK-SPIRV: Capability FPGAKernelAttributesv2INTEL
               ^
<stdin>:5:39: note: scanning from here
2 Capability FPGAKernelAttributesINTEL 
                                      ^
<stdin>:7:9: note: possible intended match here
8 Extension "SPV_INTEL_kernel_attributes" 
```

**Comment by tiwaria1** [comment]
> I'll take a look at clang-format check, you may ignore it Please take a look at the failed test: Failed Tests (1): LLVM_SPIRV :: extensions/INTEL/SPV_INTEL_kernel_attributes/register_map_interface_attribute.ll
> 

It was the missing entry in SPIRVNameMapEnum.h that was causing this. Fixed.

**Comment by tiwaria1** [comment]
What do I do about the clang-format error? Looks like it is some sort of script issue?

**Pull request merged**
